File: drills/ts/harnesses/nc82_find_median_stream.ts
import * as solution from "./solution";

const show = (v: unknown) => JSON.stringify(v) ?? "undefined";

const medians = (values: number[]) => {
  const finder = new solution.MedianFinder();
  return values.map((value) => {
    finder.addNum(value);
    return finder.findMedian();
  });
};

export function run(): [string, string, string][] {
  if (typeof solution.MedianFinder !== "function") throw new Error("__signature_mismatch__");
  return [
    ["medians of 1, 2, 3", show([1, 1.5, 2]), show(medians([1, 2, 3]))],
    ["medians of 1, 2, 3, 4, 5", show([1, 1.5, 2, 2.5, 3]), show(medians([1, 2, 3, 4, 5]))],
    ["medians arriving out of order", show([5, 3, 2, 2.5]), show(medians([5, 1, 2, 3]))],
    ["medians of negatives", show([-1, -1.5]), show(medians([-1, -2]))],
    ["median before anything is added", show(0), show(new solution.MedianFinder().findMedian())],
  ];
}

File: drills/go/harnesses/nc82_find_median_stream.go
package main

func medians(values []int) []float64 {
	finder := Constructor()
	out := []float64{}
	for _, v := range values {
		finder.AddNum(v)
		out = append(out, finder.FindMedian())
	}
	return out
}

func main() {
	run(func() []testCase {
		empty := Constructor()
		return []testCase{
			tc("medians of 1, 2, 3", []float64{1, 1.5, 2}, medians([]int{1, 2, 3})),
			tc("medians of 1, 2, 3, 4, 5", []float64{1, 1.5, 2, 2.5, 3}, medians([]int{1, 2, 3, 4, 5})),
			tc("medians arriving out of order", []float64{5, 3, 2, 2.5}, medians([]int{5, 1, 2, 3})),
			tc("medians of negatives", []float64{-1, -1.5}, medians([]int{-1, -2})),
			tc("median before anything is added", 0.0, empty.FindMedian()),
		}
	})
}
